### ISSUE TITLE
improve type resolve

### ISSUE THREAD

**Issue by ging-dev** [DEVSENSE/phptools-docs#676]
Actual:
![image](https://github.com/user-attachments/assets/e022b02d-5efb-401a-b051-bd1d2ec49066)
Expect:
![image](https://github.com/user-attachments/assets/917ddc79-fcb4-4d5b-8ae2-6e8b0af8fa8b)

**Comment by jakubmisek** [comment]
We have noticed that many existing frameworks misuse this syntax/differently :/ 

So they usually specify both `...` and `array` types (because the variable will be an array after all). Not sure what's correct.

So in case the type is `array` or `[]`, we assume, it should not be wrapped to another array again.

**Comment by ging-dev** [comment]
> We have noticed that many existing frameworks misuse this syntax/semantic :/
> 
> So they usually specify both `...` and `array` types (because the variable will be an array after all). Not sure what's correct.
> 
> So in case the type is `array` or `[]`, we assume, it should not be wrapped to another array again.
> 
> <img alt="image" width="219" src="https://private-user-images.githubusercontent.com/842150/371760497-85093e9f-8cda-41aa-a2f8-ccd2c83659fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc1NDI0MjMsIm5iZiI6MTcyNzU0MjEyMywicGF0aCI6Ii84NDIxNTAvMzcxNzYwNDk3LTg1MDkzZTlmLThjZGEtNDFhYS1hMmY4LWNjZDJjODM2NTlmZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyOFQxNjQ4NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNzRmMDk5NDdiNDk3YWVmYWU5Mjg4ZWEyZDk0M2I4OGNmZWYzNjQ4ODQ3Mzc4YzU3MTY2MDgwNzljNjE0M2Q4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.EzEySDQ8ewAxnjyNxV5I3PMAqs8aL_74-OxFEKgg0SU">



> We have noticed that many existing frameworks misuse this syntax/semantic :/
> 
> So they usually specify both `...` and `array` types (because the variable will be an array after all). Not sure what's correct.
> 
> So in case the type is `array` or `[]`, we assume, it should not be wrapped to another array again.
> 
> <img alt="image" width="219" src="https://private-user-images.githubusercontent.com/842150/371760497-85093e9f-8cda-41aa-a2f8-ccd2c83659fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc1NDI0MjMsIm5iZiI6MTcyNzU0MjEyMywicGF0aCI6Ii84NDIxNTAvMzcxNzYwNDk3LTg1MDkzZTlmLThjZGEtNDFhYS1hMmY4LWNjZDJjODM2NTlmZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkyOFQxNjQ4NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNzRmMDk5NDdiNDk3YWVmYWU5Mjg4ZWEyZDk0M2I4OGNmZWYzNjQ4ODQ3Mzc4YzU3MTY2MDgwNzljNjE0M2Q4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.EzEySDQ8ewAxnjyNxV5I3PMAqs8aL_74-OxFEKgg0SU">

Agreed, however, I think we should require users to use the correct syntax in order to synchronize with static analysis tools, some recent frameworks have started using static analysis tools:

https://github.com/illuminate/collections/blob/d91e45472a9f4be9ab21a5799308cb72172c9323/Collection.php#L1015 

I think in the example below, `string[]` is a more accurate input than `string`:
```php
<?php

require __DIR__.'/vendor/autoload.php';

/** @var \Illuminate\Support\Collection<int, string[]> */
$c = collect();

$c->push(['hello']); // Argument '1' passed to push() is expected to be of type string, string[] given
```
